kork/kork-artifacts/src/main/java/com/netflix/spinnaker/kork/artifacts/artifactstore/entities/EntityPropertyAccessor.java

```diff
@@ -0,0 +1,197 @@
+import org.springframework.expression.TypedValue;
+
+/**
+ * Used to expand manifests that live in the context, but only when a property is accessed directly.
+ *
+ * <p>SpEL expressions that just inspect the manifest will not expand it, but any access to a
+ * property within the manifest will expand the manifest and return the proper value.
+ *
+ * <p><code>
+ *     // This would appropriately expand the manifest to retrieve the value
+ *     ${ #stage('Deploy (Manifest)').context.manifests[0].metadata }
+ *     // Returns:
+ *     //  {
+ *     //    "name": "nginx",
+ *     //    "labels": {
+ *     //       "helm.sh/chart": "nginx-1.0.0",
+ *     //       "app.kubernetes.io/managed-by": "Helm",
+ *     //       "app.kubernetes.io/name": "nginx",
+ *     //       "app.kubernetes.io/instance": "nginx",
+ *     //       "app.kubernetes.io/version": "1.0.0"
+ *     //    }
+ *     //  }
+ *     // as opposed to the expression below:
+ *     ${ #stage('Deploy (Manifest)').context.manifests[0] }
+ *     // Returns:
+ *     // {
+ *     //   "customKind": false,
+ *     //   "reference": "ref://application/hash",
+ *     //   "metadata": {},
+ *     //   "name": "stored-manifest",
+ *     //   "type": "remote/manifest/base64"
+ *     // }
+ * </code>
+ */
+public class EntityPropertyAccessor implements PropertyAccessor {
+  private static final String NO_TYPE = "no-type";
+  private static final String ARTIFACT_TYPE_KEY = "type";
+  private static final String ARTIFACT_REFERENCE_KEY = "reference";
+
+  private final List<String> aggressiveExpansionKeys;
+  private final List<String> supportedTypes;
+
+  public EntityPropertyAccessor(List<String> supportedTypes, List<String> aggressiveExpansionKeys) {
+    this.supportedTypes = supportedTypes;
+    this.aggressiveExpansionKeys = aggressiveExpansionKeys;
+  }
+
+  @Override
+  public Class<?>[] getSpecificTargetClasses() {
+    return new Class[] {Map.class, Artifact.class};
+  }
+
+  @Override
+  public boolean canRead(EvaluationContext context, Object target, String name) {
+    if (ArtifactStore.getInstance() == null) {
+      return false;
+    }
+    if (this.aggressiveExpansionKeys.stream().anyMatch(k -> k.equals(name))) {
+      return true;
+    }
+
+    String type = extractType(target);
+    return this.supportedTypes.stream().anyMatch(t -> t.equals(type));
+  }
+
+  @Override
+  public TypedValue read(EvaluationContext context, Object target, String name)
+      throws AccessException {
+    if (this.aggressiveExpansionKeys.stream().anyMatch(k -> k.equals(name))) {
+      return new TypedValue(handleExpansionKey((Map) target, name));
+    }
+
+    String type = extractType(target);
+    if (this.supportedTypes.stream().noneMatch(t -> t.equals(type))) {
+      throw new ArtifactStoreInvalidTypeException(type);
+    }
+
+    if (target instanceof Map) {
+      return new TypedValue(handleMap((Map) target, name));
+    }
+    throw new ArtifactStoreInvalidStateException(
```

Comment:
If I'm following correctly, this corresponds to a 500 response code, which callers typically retry.  It doesn't make sense to retry this though, right?